bridge-worker/server.go

```diff
@@ -0,0 +1,253 @@
+
+	req, err := readSubmitAttestationRequest(r)
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err := req.Validate(); err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	events, err := s.assetsUnlockEndpoint.GetAssetsUnlockedEvents(r.Context(), req.Entry.UnlockSequence, req.Entry.UnlockSequence.AddRaw(1))
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// should have exactly one event
+	if len(events) != 1 {
+		s.writeSubmitAttestationError(w, errors.New("invalid attestation"), http.StatusBadRequest)
+		return
+	}
+
+	if !events[0].Equal(req.Entry) {
+		s.writeSubmitAttestationError(w, errors.New("invalid attestation"), http.StatusBadRequest)
+		return
+	}
+
+	// first recover the address out of the signature
+	address, err := s.recoverAddress(req.Entry, req.Signature)
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	// now validate the address is a registered validator address
+	index, err := s.mezoBridge.BridgeValidatorIDs(address)
+	if err != nil {
+		s.logger.Error("couldn't get bridge validator ID", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// if default value, then address is not a validator
+	if index == 0 {
+		s.writeSubmitAttestationError(w, errors.New("not an authorized validator"), http.StatusUnauthorized)
+		return
+	}
+
+	// then the attestation has not been confirmed yet
+	isConfirmed, err := s.mezoBridge.ConfirmedUnlocks(req.Entry.UnlockSequence.BigInt())
+	if err != nil {
+		s.logger.Error("couldn't confirm unlock", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// if already confirmed, nothing to do
+	if isConfirmed {
+		s.writeSubmitAttestationError(w, errors.New("already confirmed"), http.StatusBadRequest)
+		return
+	}
+
+	// finally, is it a valid attestation
+	attestation := toPortalAssetsUnlock(req.Entry)
+	ok, err := s.mezoBridge.ValidateAssetsUnlocked(*attestation)
+	if err != nil {
+		s.logger.Error("couldn't validate assets unlock event", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// if already confirmed, nothing to do
+	if !ok {
+		s.writeSubmitAttestationError(w, errors.New("not a valide asset unlocked event"), http.StatusBadRequest)
+		return
+	}
+
+	// now we save it into the store
+	if err := s.store.SaveAttestation(req.Entry); err != nil {
+		s.logger.Error("couldn't save assets unlock event", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	if err := s.store.SaveSignature(req.Entry.UnlockSequence, req.Signature); err != nil {
+		s.logger.Error("couldn't save signature", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
```

Comment:
I think we should aim to execute the entry and signature save as a single SQL command. Additionally, we should double-check our data was properly stored inside Supabase. This is a cloud database and we already observed weird corner cases in the past so better safe than sorry.